lib/service/postcard_service.dart

```diff
@@ -0,0 +1,305 @@
+}
+
+class PostcardServiceImpl extends PostcardService {
+  final PostcardApi _postcardApi;
+  final TezosService _tezosService;
+  final IndexerService _indexerService;
+
+  PostcardServiceImpl(
+      this._postcardApi, this._tezosService, this._indexerService);
+
+  @override
+  Future<ClaimPostCardResponse> claimEmptyPostcard(
+      ClaimPostCardRequest request) async {
+    return _postcardApi.claim(request);
+  }
+
+  @override
+  Future<ReceivePostcardResponse> receivePostcard({
+    required String shareCode,
+    required Position location,
+    required String address,
+  }) async {
+    try {
+      final body = {
+        "shareCode": shareCode,
+        "location": [location.latitude, location.longitude],
+        "address": address,
+      };
+      final respone = await _postcardApi.receive(body);
+      return respone;
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  @override
+  Future<SharePostcardResponse> sharePostcard(AssetToken asset) async {
+    final tezosService = injector<TezosService>();
+    final owner = await asset.getOwnerWallet();
+    final ownerWallet = owner?.first;
+    final addressIndex = owner?.second;
+    if (ownerWallet == null) {
+      throw Exception("Owner wallet is null");
+    }
+    final counter = asset.postcardMetadata.counter;
+    final contractAddress = asset.contractAddress ?? '';
+    final tokenId = asset.tokenId ?? '';
+    final data = [
+      TezosPack.packAddress(contractAddress),
+      TezosPack.packInteger(int.parse(tokenId)),
+      TezosPack.packInteger(counter),
+    ].toList();
+
+    final message2sign = getMessage2Sign(data);
+    final publicKey = await ownerWallet.getTezosPublicKey(index: addressIndex!);
+    final signature = await tezosService.signMessage(
+        ownerWallet, addressIndex, Uint8List.fromList(message2sign));
+    final body = {
+      "address": asset.owner,
+      "contractAddress": asset.contractAddress,
+      "signature": signature,
+      "publicKey": publicKey,
+      "counter": counter,
+    };
+    try {
+      final response = await _postcardApi.share(tokenId, body);
+
+      final deeplink = response["deeplink"];
+      return SharePostcardResponse(deeplink: deeplink);
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  @override
+  Future<AssetToken> getPostcard(String tokenId) async {
+    final request = QueryListTokensRequest(
+      ids: [tokenId],
+    );
+    final assets = await _indexerService.getNftTokens(request);
+    return assets.first;
+  }
+
+  @override
+  Future<SharedPostcardInfor> getSharedPostcardInfor(String shareCode) async {
+    try {
+      final response = await _postcardApi.claimShareCode(shareCode);
+      response["shareCode"] = shareCode;
+      final sharedPostcardInfor = SharedPostcardInfor.fromJson(response);
+      return sharedPostcardInfor;
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  @override
+  Future<bool> isReceived(String tokenId) async {
+    return false;
```

Comment:
@ppupha I saw some flow will not be reached with this hardcoded value? Is there any update needed?